apps/dashr-object-detection/app.R

```diff
@@ -0,0 +1,870 @@
+          
+          htmlDiv(
+            dict(
+              
+              htmlP(children =  "Detection Score of Most Probable Objects", className = 'plot-title'),
+              
+              dccGraph(id = 'bar-score-graph', style = dict(height = '55vh'))
+              
+              
+            )
+            
+            
+          )
+          
+          
+        )
+        
+        
+      )
+      
+    } else{
+      
+      return(dict()) # return empty dict/dict
+    }}
+)
+
+#*****************************************************************************************************************
+
+
+#Updating figures
+
+#DETECTION MODE - bar graph
+
+app$callback(
+  
+  output = dict(id = "bar-score-graph", property = "figure"),
+  
+  params = dict(input(id = "interval-detection-mode", property = "n_intervals"), 
+                
+                #State 
+                state(id = 'video-display', property='currentTime'), 
+                
+                state(id = 'dropdown-footage-selection', property = 'value'),
+                
+                state(id =  'slider-minimum-confidence-threshold', property = 'value')
+                
+  ),
+  
+  
+  
+  function(n, current_time, footage, threshold){
+    
+    if(!is.null(current_time)){
+      
+      current_frame = round(current_time * FRAMERATE)
+      
+      if((n>0)&&(current_frame>0)){
+        
+        INFO_DT = dat.dict[[footage]]$INFO_DT
+        
+        #Select the subset of the dataset
+        
+        ##Extract the score
+        frame_score <-  INFO_DT$score[(INFO_DT$frame == current_frame)]
+        
+        
+        ##Extract the class str
+        frame_string <- INFO_DT$class_str[(INFO_DT$frame == current_frame)]
+        
+        #Select only the frames above the threshold
+        
+        threshold_dec = threshold/100 #threshold in decimal
+        
+        #Filter them accordingly
+        frame_score_filtered <- frame_score[frame_score > threshold_dec]
+        frame_string_filtered <- as.character(frame_string[frame_score > threshold_dec])
+        
+        ##Save the length just in case 
+        
+        m <- 1:min(8, length(frame_score_filtered))
+        
+        frame_df <- data.table(cbind(frame_score_filtered, frame_string_filtered))
+        colnames(frame_df) <- c("Score","Class")
+        
+        #Select up to top 8 frames w.r.t scores
+        
+        frame_ordered = frame_df[order(frame_score_filtered, decreasing=TRUE),]
+        
+        top_frames <- frame_ordered[m, ]
+        
+        
+        #Add count to object names (e,g, person --> person 1, person --> person 2)
+        
+        Objects <- top_frames$Class
+        
+        set.Obj <- unique(Objects)
+        
+        Obj_count_lst <- as.list(numeric(length(set.Obj)))
+        
+        names(Obj_count_lst) <- set.Obj
+        Obj_wc = c()
+        
+        
+        
+        
+        for (Obj in Objects){
```

Comment:
we could possibly rewrite this as an lapply function. maybe something like obj_wc = lapply(Objects, function(x){Obj_wc, glue("x}  {Obj_count_lst[[x]]}")}) ?

Reply:
This requires a `left_fold` along with `lapply`, or another using `mapply` since Obj count lst is a multi-set (allowing duplicates), and we need to turn it into a disjoint union, i.e.
{a, a, b, c, b, b,b} \mapsto {(a, 1), (a,2), (b,1), (b,2), (b,3), (b,4), (c,1)}, where the index on the second component needs to be updated. I concluded a solution using `for` is more brief here.